lib/Dialect/Comb/CombFolds.cpp

```diff
@@ -1684,11 +1684,64 @@ static bool foldMuxChain(MuxOp rootMux, bool isFalseSide,
 LogicalResult MuxOp::canonicalize(MuxOp op, PatternRewriter &rewriter) {
   APInt value;
 
-  // mux(a, 1, b) -> or(a, b) for single-bit values.
-  if (matchPattern(op.trueValue(), m_RConstant(value)) &&
-      value.getBitWidth() == 1 && value.isAllOnes()) {
-    rewriter.replaceOpWithNewOp<OrOp>(op, op.cond(), op.falseValue());
-    return success();
+  if (matchPattern(op.trueValue(), m_RConstant(value))) {
+    if (value.getBitWidth() == 1) {
+      if (value.isZero()) {
+        // mux(a, 0, 1) -> ~a for single-bit values.
+        APInt value2;
+        if (matchPattern(op.falseValue(), m_RConstant(value2)) &&
+            value2.isAllOnes()) {
+          // falseValue() is known to be a single-bit 1, which we can use for
+          // the 1 in the representation of ~ using xor.
+          rewriter.replaceOpWithNewOp<XorOp>(op, op.cond(), op.falseValue());
+          return success();
+        }
```

Comment:
This isn't simplifying the single bit `mux(cond, 0, x)` case.  Can that / should that turn into an `and(~cond, x)`?

Reply:
One nice thing about handling all the 1-bit cases is that the "Else" on line 1703 would go away since this "if" would be total.  This would make it easier to understand the logic here by reducing nesting.

Reply:
Nice, I like that this makes the `else` go away. I've made this change in my latest commit. The only drawback here is that I think I can no longer rely on the trick of using the falseValue being a known one-bit `1`, meaning that I have to create a ConstantOp for the 1 in the Xor again, but that seems worth it in order to make this much more general.

I also noticed that this applies to the other conditional branch lower in the function, where we only pattern match on the falseValue being constant, so I've made the equivalent change there. This should mean that we're now able to reduce all muxes with single-bit inputs to some simpler expression with `and`, `or`, or `not`s. I've added several more FileCheck tests to double check that each of these cases is being handled properly since I don't completely trust myself to not make a bit flip error here.